diff -r fb994b2b9ab9 ChangeLog.ggx
--- a/ChangeLog.ggx	Mon Mar 10 08:00:47 2008 -0700
+++ b/ChangeLog.ggx	Mon Mar 10 08:01:10 2008 -0700
@@ -1,3 +1,18 @@ 2008-02-20  Anthony Green  <green@spinda
+2008-02-20  Anthony Green  <green@spindazzle.org>
+
+	* configure.ac: Build bfd.
+	* configure: Rebuilt.
+
+2008-02-20  Anthony Green  <green@spindazzle.org>
+
+	* configure.ac: Build ld.
+	* configure: Rebuilt.
+
+2008-02-20  Anthony Green  <green@spindazzle.org>
+
+	* configure.ac: Build binutils.
+	* configure: Rebuilt.
+
 2008-02-20  Anthony Green  <green@spindazzle.org>
 
 	* configure.ac: Build gas.
diff -r fb994b2b9ab9 configure
--- a/configure	Mon Mar 10 08:00:47 2008 -0700
+++ b/configure	Mon Mar 10 08:01:10 2008 -0700
@@ -2327,7 +2327,7 @@ case "${target}" in
     ;;
   ggx-*-*)
     noconfigdirs="$noconfigdirs ${libgcj} gdb sim"
-    noconfigdirs="$noconfigdirs binutils ld gprof"
+    noconfigdirs="$noconfigdirs ld gprof"
     ;;
   h8300*-*-*)
     noconfigdirs="$noconfigdirs target-libgloss ${libgcj}"
diff -r fb994b2b9ab9 configure.ac
--- a/configure.ac	Mon Mar 10 08:00:47 2008 -0700
+++ b/configure.ac	Mon Mar 10 08:01:10 2008 -0700
@@ -605,7 +605,7 @@ case "${target}" in
     ;;
   ggx-*-*)
     noconfigdirs="$noconfigdirs ${libgcj} gdb sim"
-    noconfigdirs="$noconfigdirs binutils ld gprof"
+    noconfigdirs="$noconfigdirs ld gprof"
     ;;
   h8300*-*-*)
     noconfigdirs="$noconfigdirs target-libgloss ${libgcj}"
